test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -2345,32 +2355,61 @@ private static class ESRestTestCaseHistoricalFeatures implements FeatureSpecific
+            }
 
-                String[] metadataFiles = metadataPath.split(System.getProperty("path.separator"));
-                for (String metadataFile : metadataFiles) {
-                    try (
-                        InputStream in = Files.newInputStream(PathUtils.get(metadataFile));
-                        XContentParser parser = JsonXContent.jsonXContent.createParser(XContentParserConfiguration.EMPTY, in)
-                    ) {
-                        for (Map.Entry<String, String> entry : parser.mapStrings().entrySet()) {
-                            historicalFeaturesMap.put(new NodeFeature(entry.getKey()), Version.fromString(entry.getValue()));
+            return historicalFeatures;
+        }
+
+        static Map<NodeFeature, Version> loadHistoricalFeatures(String metadataPath) {
+            Map<NodeFeature, Version> historicalFeaturesMap = new HashMap<>();
+            String[] metadataFiles = metadataPath.split(System.getProperty("path.separator"));
+            for (String metadataFile : metadataFiles) {
+                try (
+                    InputStream in = Files.newInputStream(PathUtils.get(metadataFile));
+                    XContentParser parser = JsonXContent.jsonXContent.createParser(XContentParserConfiguration.EMPTY, in)
+                ) {
+                    Map<String, Object> parsedMap = parser.map();
+                    if (parsedMap.get("historical_features") instanceof Map<?, ?> parsedHistoricalFeatures) {
+                        for (var entry : parsedHistoricalFeatures.entrySet()) {
+                            historicalFeaturesMap.put(
+                                new NodeFeature((String) entry.getKey()),
+                                Version.fromString((String) entry.getValue())
+                            );
                         }
-                    } catch (IOException e) {
-                        throw new UncheckedIOException(e);
                     }
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
                 }
-
-                historicalFeatures = Collections.unmodifiableMap(historicalFeaturesMap);
             }
+            return historicalFeaturesMap;
+        }
 
-            return historicalFeatures;
+        static Set<String> loadFeatureNames(String metadataPath) {
+            Set<String> featureNames = new HashSet<>();
+            String[] metadataFiles = metadataPath.split(System.getProperty("path.separator"));
+            for (String metadataFile : metadataFiles) {
+                try (
+                    InputStream in = Files.newInputStream(PathUtils.get(metadataFile));
+                    XContentParser parser = JsonXContent.jsonXContent.createParser(XContentParserConfiguration.EMPTY, in)
+                ) {
+                    Map<String, Object> parsedMap = parser.map();
+                    if (parsedMap.get("feature_names") instanceof ArrayList<?> parsedFeatureNames) {
```

Comment:
Shouldn't be checking against concrete types - should be `Collection<?>` instead